### PR TITLE
[16.0][ADD] project_task_parent_due_auto: Updates parent task due date when child due date changes.

### PR DESCRIPTION
**Pull request by david-banon-tecnativa**
@tecnativa @pedrobaeza @juancarlosonate-tecnativa

TT55845

### CODE REVIEW

File: project_task_parent_due_auto/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Copyright 2024 Tecnativa Carolina Fernandez
```
Comment (juancarlosonate-tecnativa) [inline project_task_parent_due_auto/__manifest__.py]:
```suggestion
# Copyright 2025 Tecnativa David Bañón
```

File: project_task_parent_due_auto/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+{
+    "name": "Project task parent due date auto",
+    "version": "16.0.1.0.0",
+    "category": "Project Management",
+    "summary": "Recalcualts parent taskd ue date when child changes",
```
Comment (juancarlosonate-tecnativa) [inline project_task_parent_due_auto/__manifest__.py]:
```suggestion
    "summary": "Recalculates parent task's due date when child task changes",
```

**Review by juancarlosonate-tecnativa — APPROVED** [review]
LGTM!

**Comment by david-banon-tecnativa** [comment]
Fixed, ty @juancarlosonate-tecnativa

File: project_task_parent_due_auto/models/project_task.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+                    parent.date_deadline = task.date_deadline
+                else:
+                    closest_task = min(
+                        parent.child_ids, key=lambda t: t.date_deadline or date.max
+                    )
+                    if (
+                        closest_task.date_deadline
+                        and parent.date_deadline != closest_task.date_deadline
+                    ):
+                        parent.date_deadline = closest_task.date_deadline
```
Comment (pedrobaeza) [inline project_task_parent_due_auto/models/project_task.py]: Please don't over-complicate:

```suggestion
                deadline = min(parent.child_ids.mapped("date_deadline"))
                if deadline and parent.date_deadline != deadline
                    parent.date_deadline = deadline
```

File: project_task_parent_due_auto/readme/CONTRIBUTORS.rst [hunk 4]
```diff
@@ -0,0 +1,4 @@
+* `Tecnativa <https://www.tecnativa.com>`_:
+
+    > - Pedro M. Baeza
+    > - David Bañón Gil
```
Comment (pedrobaeza) [inline project_task_parent_due_auto/readme/CONTRIBUTORS.rst]: Incorrect syntax:

```suggestion
    - Pedro M. Baeza
    - David Bañón Gil
```

File: project_task_parent_due_auto/models/project_task.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+class ProjectTask(models.Model):
+    _inherit = "project.task"
+
+    @api.model
+    def create(self, values):
+        tasks = super().create(values)
+        if "date_deadline" in values:
+            tasks.update_parent_task_dates()
+        return tasks
```
Comment (pedrobaeza) [inline project_task_parent_due_auto/models/project_task.py]: You are decreasing the performance not doing it multi-batch and mixing some concepts:

```suggestion
    @api.model_create_multi
    def create(self, vals_list):
        tasks = super().create(vals_list)
        for vals, task in zip(vals_list, tasks):
            if "date_deadline" in vals:
                task.update_parent_task_dates()
        return tasks
```

File: project_task_parent_due_auto/tests/test_parent_due_auto.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                    "name": "Test Task 3",
+                    "user_ids": cls.user_projectuser,
+                    "project_id": cls.project_pigs.id,
+                }
+            )
+        )
+
+    def setUp(self):
+        super().setUp()
+        self.task_2.parent_id = self.task_1
```
Comment (pedrobaeza) [inline project_task_parent_due_auto/tests/test_parent_due_auto.py]: Do everything in `setUpClass`.

File: project_task_parent_due_auto/tests/test_parent_due_auto.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+    def test_setting_parent_due_date_does_not_change_child(self):
+        self.task_1.write({"date_deadline": self.date_1})
+        self.assertFalse(
+            self.task_2.date_deadline,
+            "Updaitng parent deadline should not update child",
```
Comment (pedrobaeza) [inline project_task_parent_due_auto/tests/test_parent_due_auto.py]:
```suggestion
            "Updating parent deadline should not update child",
```

File: project_task_parent_due_auto/tests/test_parent_due_auto.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+from odoo import fields
+
+from odoo.addons.project.tests.test_project_base import TestProjectCommon
+
+
+class TestParentDueAuto(TestProjectCommon):
+    @classmethod
+    def setUpClass(cls):
+        super(TestParentDueAuto, cls).setUpClass()
```
Comment (pedrobaeza) [inline project_task_parent_due_auto/tests/test_parent_due_auto.py]:
```suggestion
        super().setUpClass()
```

File: project_task_parent_due_auto/tests/test_parent_due_auto.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
+from dateutil import relativedelta
+
+from odoo import fields
+
+from odoo.addons.project.tests.test_project_base import TestProjectCommon
+
+
+class TestParentDueAuto(TestProjectCommon):
```
Comment (pedrobaeza) [inline project_task_parent_due_auto/tests/test_parent_due_auto.py]: I think this is too much overhead for simple tests, don't you think. Just create here a project and a couple of tasks for testing this feature.

File: project_task_parent_due_auto/tests/test_parent_due_auto.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+        self.env["project.task"].with_context(mail_create_nolog=True).create(
+            {
+                "name": "Child Task",
+                "user_ids": self.user_projectuser,
+                "project_id": self.project_pigs.id,
+                "parent_id": self.task_1.id,
+                "date_deadline": self.date_10,
+            }
+        )
+
```
Comment (pedrobaeza) [inline project_task_parent_due_auto/tests/test_parent_due_auto.py]: Don't put empty lines inside a method.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: project_task_parent_due_auto/tests/test_parent_due_auto.py (hunk 9 above)
Comment (david-banon-tecnativa) [reply project_task_parent_due_auto/tests/test_parent_due_auto.py]: They are too much overhead, I wrote them to test edge cases I imagined during development but guessed it didn't hurt to leave them, I will simplify the tests as much i can.

Comment (pedrobaeza) [reply project_task_parent_due_auto/tests/test_parent_due_auto.py]: I mean to inherit from `TestProjectCommon` instead of just `TransactionCase`.

Comment (david-banon-tecnativa) [reply project_task_parent_due_auto/tests/test_parent_due_auto.py]: What do you mean by overhead in this case? We don't need many of the things `TestProjectCommon` does, but i don't see how it impacts negatively to use it instead of creating projects and tasks on our own.

Comment (pedrobaeza) [reply project_task_parent_due_auto/tests/test_parent_due_auto.py]: It impacts in more test time creating that extra records that we are not using, and at the end, this means more total test time on each CI, deployment, etc.

Comment (david-banon-tecnativa) [reply project_task_parent_due_auto/tests/test_parent_due_auto.py]: Makes sense, I'll fix this and keep it in mind.

**Comment by david-banon-tecnativa** [comment]
@pedrobaeza all changes have been taken care of

File: project_task_parent_due_auto/models/project_task.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+    def update_parent_task_dates(self):
+        for task in self:
+            parent = task.parent_id
+            if parent:
+                deadline = min(
+                    parent.child_ids.mapped(lambda t: t.date_deadline or date.max)
+                )
+                if deadline == date.max:
+                    parent.date_deadline = False
```
Comment (pedrobaeza) [inline project_task_parent_due_auto/models/project_task.py]:
```suggestion
                children = parent.child_ids.filtered("date_deadline")
                deadline = children and min(children.mapped("date_deadline"))
```

Comment (pedrobaeza) [reply project_task_parent_due_auto/models/project_task.py]: Modified for the case where no task has deadline.

Comment (david-banon-tecnativa) [reply project_task_parent_due_auto/models/project_task.py]: This changes the behavior to keeping the parent deadline when the child deadline is removed. To keep functionality, these are the options i see:
```python
# Same as last commit
deadline = min(
    parent.child_ids.mapped(lambda t: t.date_deadline or date.max)
)
if deadline == date.max:
    parent.date_deadline = False
elif deadline and parent.date_deadline != deadline:
    parent.date_deadline = deadline
``` 
Or
```python
# The changes you reviewed
children = parent.child_ids.filtered("date_deadline")
deadline = bool(children) and min(children.mapped("date_deadline"))
# If now might be unnecessary
if parent.date_deadline != deadline:
    parent.date_deadline = deadline
```

Comment (pedrobaeza) [reply project_task_parent_due_auto/models/project_task.py]: But that's not a requirement right now in the customer needs.

Comment (david-banon-tecnativa) [reply project_task_parent_due_auto/models/project_task.py]: I keep overthinking requirements, need to stop doing that, changes have been implemented and the test has been inverted.

File: project_task_parent_due_auto/tests/test_parent_due_auto.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        self.task_2.write({"date_deadline": self.date_5})
+        self.assertEqual(
+            self.task_1.date_deadline,
+            self.date_5,
+            "Updating child deadline should update parent date",
+        )
+
+    def test_removing_child_due_date_removes_from_parent(self):
+        self.task_2.write({"date_deadline": self.date_5})
+        # import wdb; wdb.set_trace()
```
Comment (pedrobaeza) [inline project_task_parent_due_auto/tests/test_parent_due_auto.py]: You have forgotten this.

Comment (david-banon-tecnativa) [reply project_task_parent_due_auto/tests/test_parent_due_auto.py]: Removed

**Review by pedrobaeza — APPROVED** [review]
/ocabot merge nobump
